include/ginkgo/core/base/exception.hpp

```diff
@@ -173,6 +173,64 @@ class NotSupported : public Error {
 };
 
 
+/**
+ * MpiError is thrown when a MPI routine throws a non-zero error code.
+ */
+class MpiError : public Error {
+public:
+    /**
+     * Initializes a MPI error.
+     * @param file The name of the offending source file
+     * @param line The source code line number where the error occurred
+     * @param func The name of the MPI routine that failed
+     * @param error_code The resulting MPI error code
+     */
+    MpiError(const std::string& file, int line, const std::string& func,
+             int64 error_code)
+        : Error(file, line, func + ": " + get_error(error_code))
+    {}
+
+private:
+    static std::string get_error(int64 error_code);
+};
+
+
+/**
+ * MpiError is thrown when a MPI has already been initialized.
+ * MPI_Init can only be called once in a program.
+ */
+class MpiInitialized : public Error {
+public:
+    /**
+     * Initializes a MPI error.
+     * @param file The name of the offending source file
+     * @param line The source code line number where the error occurred
+     * @param func The name of the MPI routine that failed
+     */
+    MpiInitialized(const std::string& file, int line, const std::string& func)
+        : Error(file, line, func)
+    {}
+};
+
+
+/**
+ * MpiError is thrown when a MPI has already been finalized.
+ * Any MPI routines that are called after calling MPI_Finalize throw this error.
+ */
+class MpiFinalized : public Error {
+public:
+    /**
+     * Initializes a MPI error.
+     * @param file The name of the offending source file
+     * @param line The source code line number where the error occurred
+     * @param func The name of the MPI routine that failed
+     */
+    MpiFinalized(const std::string& file, int line, const std::string& func)
+        : Error(file, line, func)
+    {}
```

Comment:
This should be fixed now with a singleton

Reply:
I think regardless of how we deal with Init/Finalize, we don't really need to guard against them being called twice, since all MPI implementations I tested complained very loudly about this and aborted the program straight away.